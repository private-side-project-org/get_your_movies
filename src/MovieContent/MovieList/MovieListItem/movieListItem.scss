@use "assets/variables" as *;

.moviesListItem {
  &-container {
    display: flex;
    flex-direction: column;
    max-height: 300px;
    padding: 8px;
    margin-bottom: 16px;
    margin-right: 8px;
    border: 1px solid gray;
    border-radius: 4px;
    gap: 8px;

    animation-name: moveRight;
    animation-duration: 0.5s;
    animation-timing-function: ease-in;

    > p {
      display: -webkit-box;
      -webkit-line-clamp: 3;
      -webkit-box-orient: vertical;
      overflow: hidden;
    }

    &:hover {
      background-color: lightgray;
      > img {
        opacity: 0.8;
      }
      cursor: pointer;
      box-shadow: 5px 5px 5px;
    }

    > img:not(.moviesListItem-no-image) {
      width: 100%;
      max-width: fit-content;
    }
  }

  &-no-image {
    width: 100px;
  }
}
