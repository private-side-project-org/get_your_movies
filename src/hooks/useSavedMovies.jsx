import React, { createContext, useContext, useState } from "react";
import PropTypes from "prop-types";
const SavedMovieContext = createContext({});

const savedMovieContextValues = () => {
  const [favoriteMovieList, setFavoriteMovieList] = useState(
    JSON.parse(localStorage.getItem("movies"))
  );
  const getFavoriteMovie = (movie) =>
    favoriteMovieList.find((favMovie) => favMovie.id === movie.id);

  return {
    favoriteMovieList,
    setFavoriteMovieList,
    getFavoriteMovie,
  };
};

export const SavedMoviesProvider = ({ children }) => {
  return (
    <SavedMovieContext.Provider value={savedMovieContextValues()}>
      {children}
    </SavedMovieContext.Provider>
  );
};

SavedMoviesProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default () => useContext(SavedMovieContext);
