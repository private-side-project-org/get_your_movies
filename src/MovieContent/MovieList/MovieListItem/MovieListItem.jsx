import React from "react";
import PropTypes from "prop-types";
import getMovieImagePath from "utils/getMovieImagePath";
import useSavedMovies from "hooks/useSavedMovies";

import "./movieListItem.scss";

const noImageAvailable = require("assets/icons/camera-reel.svg");

const propTypes = {
  movie: PropTypes.shape({
    id: PropTypes.number,
    original_title: PropTypes.string,
    overview: PropTypes.string,
    backdrop_path: PropTypes.string,
  }).isRequired,
  onSetSelectedMovie: PropTypes.func.isRequired,
  isFirstMovie: PropTypes.bool.isRequired,
  selectedTab: PropTypes.string.isRequired,
};

const MovieListItem = ({ movie, onSetSelectedMovie, isFirstMovie }) => {
  const { getFavoriteMovie } = useSavedMovies();
  const { original_title, overview, backdrop_path } = movie;
  return (
    <div
      className="moviesListItem-container"
      onClick={() => {
        const favoriteMovie = getFavoriteMovie(movie);

        // set favoriteMovie to prevent extra api call if exists
        const movieToBeSet = favoriteMovie || { id: movie.id };
        onSetSelectedMovie(movieToBeSet);
      }}
      // set id #top to get user to go top when click `top` button
      id={isFirstMovie ? "top" : ""}
    >
      <h3>{original_title}</h3>
      {backdrop_path ? (
        <img src={getMovieImagePath(backdrop_path, 300)} alt="movie_label" />
      ) : (
        <img
          src={noImageAvailable}
          alt="no_image_available"
          className="moviesListItem-no-image"
        />
      )}
      <p>{overview}</p>
    </div>
  );
};

MovieListItem.propTypes = propTypes;

export default MovieListItem;
