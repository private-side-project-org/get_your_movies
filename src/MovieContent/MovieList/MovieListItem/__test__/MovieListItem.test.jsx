import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import MovieListItem from "../MovieListItem";
import { SavedMovieProvider } from "hooks/useSavedMovies";

const onSetSelectedMovie = () => null;
const movie = {
  original_title: "test title",
  overview: "this is test overview",
  backdrop_path: "abcde12345",
};

let renderedMovieListItem;

beforeEach(() => {
  renderedMovieListItem = render(
    <MovieListItem
      movie={movie}
      onSetSelectedMovie={onSetSelectedMovie}
      selectedTab="search"
      isFirstMovie={false}
    />,
    { wrapper: SavedMovieProvider }
  );
});

describe("MovieListItem component test", () => {
  it("gets props and generate snapshot", () => {
    expect(renderedMovieListItem.baseElement).toMatchSnapshot();
  });

  it("get correct title, overview", () => {
    const { getByText } = renderedMovieListItem;

    getByText("test title");
    getByText("this is test overview");
  });

  it("renders, get proper image src", async () => {
    const { findByAltText } = renderedMovieListItem;
    const targetImg = await findByAltText("movie_label");

    expect(targetImg.src).toContain(
      "https://image.tmdb.org/t/p/w300abcde12345"
    );
  });

  it("fire click event", () => {
    const targetElem = screen.getByTestId("set-movie-button");

    fireEvent(
      targetElem,
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );
  });
});
