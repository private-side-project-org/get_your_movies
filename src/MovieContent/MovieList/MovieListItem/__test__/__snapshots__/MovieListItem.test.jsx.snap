// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieListItem component test gets props and generate snapshot 1`] = `
<body>
  <div>
    <div
      class="moviesListItem-container"
      data-testid="set-movie-button"
      id=""
    >
      <h3>
        test title
      </h3>
      <img
        alt="movie_label"
        src="https://image.tmdb.org/t/p/w300abcde12345"
      />
      <p>
        this is test overview
      </p>
    </div>
  </div>
</body>
`;
